src/material/tree/node.ts

```diff
@@ -25,36 +25,106 @@ import {
   booleanAttribute,
   numberAttribute,
 } from '@angular/core';
+import {CanDisable, HasTabIndex} from '@angular/material/core';
+import {BooleanInput, coerceBooleanProperty, coerceNumberProperty} from '@angular/cdk/coercion';
+import {LegacyTreeKeyManager, TreeKeyManagerItem, TreeKeyManagerStrategy} from '@angular/cdk/a11y';
+
+/**
+ * Determinte if argument TreeKeyManager is the LegacyTreeKeyManager. This function is safe to use with SSR.
+ */
+function isLegacyTreeKeyManager<T extends TreeKeyManagerItem>(
+  keyManager: TreeKeyManagerStrategy<T>,
+): keyManager is LegacyTreeKeyManager<T> {
+  return !!(keyManager as any)._isLegacyTreeKeyManager;
+}
 
 /**
  * Wrapper for the CdkTree node with Material design styles.
  */
 @Directive({
   selector: 'mat-tree-node',
   exportAs: 'matTreeNode',
+  outputs: ['activation', 'expandedChange'],
   providers: [{provide: CdkTreeNode, useExisting: MatTreeNode}],
   host: {
     'class': 'mat-tree-node',
+    '[attr.aria-expanded]': '_getAriaExpanded()',
+    '[attr.aria-level]': 'level + 1',
+    '[attr.aria-posinset]': '_getPositionInSet()',
+    '[attr.aria-setsize]': '_getSetSize()',
+    '(click)': '_focusItem()',
+    'tabindex': '_getTabindexAttribute()',
   },
 })
 export class MatTreeNode<T, K = T> extends CdkTreeNode<T, K> implements OnInit, OnDestroy {
-  /** Whether the node is disabled. */
-  @Input({transform: booleanAttribute})
-  disabled: boolean = false;
+  /**
+   * The tabindex of the tree node.
+   *
+   * @deprecated By default MatTreeNode manages focus using TreeKeyManager instead of tabIndex.
+   *   Recommend to avoid setting tabIndex directly to prevent TreeKeyManager form getting into
+   *   an unexpected state. Tabindex to be removed in a future version.
+   * @breaking-change 19.0.0 Remove this attribute.
+   */
 
-  /** Tabindex of the node. */
   @Input({
     transform: (value: unknown) => (value == null ? 0 : numberAttribute(value)),
   })
-  tabIndex: number;
+  get tabIndex(): number {
+    return this.isDisabled ? -1 : this._tabIndex;
+  }
+  set tabIndex(value: number) {
+    // If the specified tabIndex value is null or undefined, fall back to the default value.
+    this._tabIndex = value;
+  }
+  private _tabIndex: number;
+
+  /**
+   * The default tabindex of the tree node.
+   *
+   * @deprecated By default MatTreeNode manages focus using TreeKeyManager instead of tabIndex.
+   *   Recommend to avoid setting tabIndex directly to prevent TreeKeyManager form getting into
+   *   an unexpected state. Tabindex to be removed in a future version.
+   * @breaking-change 19.0.0 Remove this attribute.
+   */
+  defaultTabIndex = 0;
```

Comment:
Why is this being introduced as deprecated instead of just private/internal?

Reply:
defaultTabIndex has been here all along. this PR doesn't actually introduce `defaultTabIndex`. That's because before it was coming from `hasTabindex` mixin. This change removes the mixin so we can mark the members as deprecated and change the logic of how tabindex is set – by default MatTreeNode ignores the tabindex Input provided and uses the key manager instead.